installer/windows/Storj/CustomAction.cs

```diff
@@ -8,195 +8,277 @@ namespace Storj
-                return ActionResult.Success;
+                throw new ArgumentException("You must select a storage folder.");
             }
 
-            DirectoryInfo dir = new DirectoryInfo(identityDir);
+            DirectoryInfo dir = new DirectoryInfo(storageDir);
             DriveInfo drive = new DriveInfo(dir.Root.FullName);
 
             // TODO: Find a way to calculate the available free space + total size of existing pieces
             if (drive.TotalSize < MinFreeSpace)
             {
-                session["STORJ_STORAGEDIR_VALID"] = string.Format("The selected drive '{0}' has only {1:0.##} GB disk size. The minimum required is 550 GB.",
-                    drive.Name, decimal.Divide(drive.TotalSize, GB));
-                return ActionResult.Success;
+                throw new ArgumentException(string.Format("The selected drive '{0}' has only {1:0.##} GB disk size. The minimum required is 550 GB.",
+                    drive.Name, decimal.Divide(drive.TotalSize, GB)));
             }
-
-            // Storage dir is valid
-            session["STORJ_STORAGEDIR_VALID"] = "1";
-            return ActionResult.Success;
         }
 
-        [CustomAction]
-        public static ActionResult ValidateStorage(Session session)
+        public static void ValidateStorage(string storageStr, string storageDir)
         {
-            string storageStr = session["STORJ_STORAGE"];
-
             if (string.IsNullOrEmpty(storageStr))
             {
-                session["STORJ_STORAGE_VALID"] = "The value cannot be empty.";
-                return ActionResult.Success;
+                throw new ArgumentException("The value cannot be empty.");
             }
 
             if (!double.TryParse(storageStr, NumberStyles.Number, CultureInfo.CreateSpecificCulture("en-US"), out double storage))
             {
-                session["STORJ_STORAGE_VALID"] = string.Format("'{0}' is not a valid number.", storageStr);
-                return ActionResult.Success;
+                throw new ArgumentException(string.Format("'{0}' is not a valid number.", storageStr));
             }
 
-            if (storage < 0.5) {
-                session["STORJ_STORAGE_VALID"] = "The allocated disk space cannot be less than 0.5 TB.";
-                return ActionResult.Success;
+            if (storage < 0.5)
+            {
+                throw new ArgumentException("The allocated disk space cannot be less than 0.5 TB.");
+            }
+
+            if (storageDir == null)
+            {
+                throw new ArgumentException("The storage directory cannot be null");
             }
 
-            DirectoryInfo dir = new DirectoryInfo(session["STORAGEDIR"]);
+            DirectoryInfo dir = new DirectoryInfo(storageDir);
             DriveInfo drive = new DriveInfo(dir.Root.FullName);
             long storagePlusOverhead = Convert.ToInt64(storage * 1.1 * TB);
 
```

Comment:
We should add a test with an input of 10000000. This conversion makes the installer crash.

Reply:
Good catch!

Reply:
Test added.